### PR TITLE
TokenStore: Added cleanup endpoint

### PR DESCRIPTION
**Pull request by vishalnayak**

### CODE REVIEW

File: vault/token_store.go [hunk 1]
```diff
@@ -3,12 +3,14 @@ package vault
 import (
 	"encoding/json"
 	"fmt"
+	"log"
```
Comment (jefferai) [inline vault/token_store.go]: A reminder to remove the log statements before merging.

File: vault/token_store.go [hunk 2]
```diff
@@ -1010,6 +1023,84 @@ func (ts *TokenStore) lookupBySaltedAccessor(saltedAccessor string) (accessorEnt
...
+	// a valid one. If not, delete the leases associated with that token
+	// and delete the accessor as well.
+	for _, saltedAccessor := range saltedAccessorList {
+		accessorEntry, err := ts.lookupBySaltedAccessor(saltedAccessor)
+		if err != nil {
+			cleanupErrors = multierror.Append(cleanupErrors, fmt.Errorf("failed to read the accessor entry: %v", err))
+			continue
+		}
+
+		if accessorEntry.TokenID == "" {
```
Comment (jefferai) [inline vault/token_store.go]: Please add a comment above to make it clear why the rest is being skipped -- that this entry itself is bogus.

File: vault/token_store.go [hunk 3]
```diff
@@ -1010,6 +1023,84 @@ func (ts *TokenStore) lookupBySaltedAccessor(saltedAccessor string) (accessorEnt
...
+		te, err := ts.lookupSalted(ts.SaltID(accessorEntry.TokenID))
+
+		// If token entry is not fetched for whatever reason, assume
+		// that the token is no more valid and conclude that accessor
+		// for this token should not exist along with the leases
+		// associated with the token. Also, in case token entry exists
+		// and if NumUses in the token entry is marked as -3, it a hint
+		// that the previous revocation operation failed. In that case,
+		// attempt a retry operation.
+		if err != nil || te == nil || (te != nil && te.NumUses == -3) {
```
Comment (jefferai) [inline vault/token_store.go]: Note that there are const values now -- this would be `tokenRevocationFailed`

File: vault/token_store.go [hunk 4]
```diff
@@ -1010,6 +1023,84 @@ func (ts *TokenStore) lookupBySaltedAccessor(saltedAccessor string) (accessorEnt
...
+			// If deletion of accessor fails, move on to the next
+			// item since this is just a best-effort operation
+			err = ts.view.Delete(accessorPrefix + saltedAccessor)
+			if err != nil {
+				cleanupErrors = multierror.Append(cleanupErrors, fmt.Errorf("failed to delete the accessor entry: %v", err))
+				continue
+			}
+		}
+
+		te, err := ts.lookupSalted(ts.SaltID(accessorEntry.TokenID))
```
Comment (jefferai) [inline vault/token_store.go]: You'll need to make this `te, err := ts.lookupSalted(ts.SaltID(accessorEntry.TokenID), true)`

File: vault/token_store.go [hunk 5]
```diff
@@ -1010,6 +1023,84 @@ func (ts *TokenStore) lookupBySaltedAccessor(saltedAccessor string) (accessorEnt
...
+			// with -3 again, so that it can be retried. Since it
+			// will be handled during the next invocation of this
+			// endpoint, ignoring the error case here.
+
+			if err == nil {
+				// If deletion of accessor fails, move on to the next
+				// item since this is just a best-effort operation
+				err = ts.view.Delete(accessorPrefix + saltedAccessor)
+				if err != nil {
+					cleanupErrors = multierror.Append(cleanupErrors, fmt.Errorf("failed to delete the accessor entry: %v", err))
```
Comment (jefferai) [inline vault/token_store.go]: s/to delete the accessor/to delete accessor/

File: vault/token_store.go [hunk 6]
```diff
@@ -2024,6 +2115,15 @@ func (ts *TokenStore) tokenStoreRoleCreateUpdate(
 }
 
 const (
+	tokenCleanupHelp = `
```
Comment (jefferai) [inline vault/token_store.go]: I'd say "This endpoint performs cleanup tasks that can be run if certain error conditions have occurred."

File: vault/token_store.go [hunk 7]
```diff
@@ -2024,6 +2115,15 @@ func (ts *TokenStore) tokenStoreRoleCreateUpdate(
...
+This endpoint cleans up the storage and lease entries.
+`
+	tokenCleanupDesc = `
+This is provided to be able to cleanup accessor storage entries that were
```
Comment (jefferai) [inline vault/token_store.go]: I'd leave it less specific as we may add things later. Something like "This endpoint performs cleanup tasks that can be run to clean up token and lease entries after certain error conditions. Usually running this is not necessary, and is only required if upgrade notes or support personnel suggest it."

File: vault/token_store.go [hunk 8]
```diff
@@ -1010,6 +1022,85 @@ func (ts *TokenStore) lookupBySaltedAccessor(saltedAccessor string) (accessorEnt
...
+		te, err := ts.lookupSalted(ts.SaltID(accessorEntry.TokenID))
+
+		// If token entry is not fetched for whatever reason, assume
+		// that the token is no more valid and conclude that accessor
+		// for this token should not exist, the leases associated with
+		// the token should not exist as well. Also, in case token
+		// entry exists, and if NumUses in the token entry is marked as
+		// -3, it a hint that the previous revocation operation had
+		// failed. In that case, attempt a retry operation.
+		if err != nil || te == nil || (te != nil && te.NumUses == -3) {
```
Comment (jefferai) [inline vault/token_store.go]: Are you sure that the right thing to do in an *error* case is perform this revocation? An error case is likely to be some kind of underlying storage error, I think; not finding the entry should simply return nil.

Comment (vishalnayak) [inline vault/token_store.go]: Sounds good.

Comment (vishalnayak) [inline vault/token_store.go]: Yeah, lets skip the error case and consider only the `te == nil` case here.

**Pull request merged**